const-generic-const-fn-bounds.md

```diff
@@ -0,0 +1,496 @@
+}
+// ...desugars to...
+trait V {
+    const fn foo(C) -> D;
+    const fn bar(E) -> F;
+}
+```
+
+## `?const` modifiers in trait methods
+
+This RFC does not touch `trait` methods at all, all traits are defined as they would be defined
+without `const` functions existing. A future extension could allow
+
+```rust
+trait Foo {
+    fn a<T: ?const Bar>() -> i32;
+}
+```
+
+which does not force `impl const Foo for Type` to now require passing a `T` with an `impl const Bar`
+to the `a` method.
+
+## `const` function pointers
+
+```rust
+const fn foo(f: fn() -> i32) -> i32 {
+    f()
+}
+```
+
+is currently illegal. While we can change the language to allow this feature, two questions make
+themselves known:
+
+1. fn pointers in constants
+
+    ```rust
+    const F: fn() -> i32 = ...;
+    ```
+
+    is already legal in Rust today, even though the `F` doesn't need to be a `const` function.
+
+2. Opt out bounds are ugly
+
+    I don't think it's either intuitive nor readable to write the following
+
+    ```rust
+    const fn foo(f: ?const fn() -> i32) -> i32 {
+        // not allowed to call `f` here, because we can't guarantee that it points to a `const fn`
+    }
+    ```
+
+Thus it seems useful to prefix function pointers to `const` functions with `const`:
+
+```rust
+const fn foo(f: const fn() -> i32) -> i32 {
+    f()
+}
+const fn bar(f: fn() -> i32) -> i32 {
+    f() // ERROR
+}
+```
+
+This opens up the curious situation of `const` function pointers in non-const functions:
+
+```rust
+fn foo(f: const fn() -> i32) -> i32 {
+    f()
+}
+```
+
+Which is useless except for ensuring some sense of "purity" of the function pointer ensuring that
+subsequent calls will only modify global state if passed in via arguments.
+
+## explicit `const` bounds
+
+`const` on the bounds (e.g. `T: const Trait`) requires an `impl const Trait` for any types used to
+replace `T`. This allows `const` trait bounds on any (even non-const) functions, e.g. in
+
+```rust
+fn foo<T: const Bar>() -> i32 {
+    const FOO: i32 = T::bar();
+    FOO
+}
+```
+
+Which, once `const` items and array lengths inside of functions can make use of the generics of
+the function, would allow the above function to actually exist.
+
+## `const` default method bodies
+
+Trait methods can have default bodies for methods that are used if the method is not mentioned
+in an `impl`. This has several uses, most notably
+
+* reducing code repetition between impls that are all the same
+* adding new methods is not a breaking change if they also have a default body
+
+In order to keep both advantages in the presence of `impl const`s, we need a way to declare the
+method default body as being `const`. The author of this RFC considers prepending the default body's
+method signature with `const` to be the most intuitive syntax.
+
+```rust
+trait Foo {
+    const fn bar() {}
+}
+```
+
+While this conflicts with other future work ideas like `const` trait methods or `const trait` declarations,
```

Comment:
As far as I can tell, this doesn't conflict: these two syntaxes complement each other. Allowing default `const` method definitions is strong motivation for allowing `const` on methods (and `const trait` in general) from the beginning in my view.

Reply:
How would you declare a method whose default body is `const`, but the method is not required to have a `const` body in implementations? This question is the conflict that I mean and I should clarify this.

Reply:
This is a good question: it comes from the distinction between traits as signatures versus definitions. Can we use specialisation directly for this? I.e. the trait signature is always `const`-free to make it forwards compatible with adding this sugar? So if you want a `const` default implementation that doesn't require implementers to `const`-implement it, then you provide `default const fn` in an `impl`, but don't specify it `const` in the trait signature itself?

Reply:
Uh... specialization doesn't really seem appropriate here. It's a completely different scheme. We could also go with an attribute

```rust
trait Foo {
    #[default_method_body_is_const]
    fn foo() {}
}
```

I do see how

```rust
trait Foo {
    const fn foo() {}
}
```

could give the image of the function being const and not just the default body being so. So some other scheme should be chosen regardless of whether `const` methods are a thing we'll ever want.

Reply:
> Uh... specialization doesn't really seem appropriate here. It's a completely different scheme.

It's supposed to be a desugaring of default trait implementations. (But perhaps not intuitively from a user's POV).

> We could also go with an attribute

I would be happy going with an attribute for the proposal, but possibly leaving it as something to revisit for stabilisation?